cluster/manager.go

```diff
@@ -738,77 +720,97 @@ func (c *ClusterManager) startListeners(db kvdb.Kvdb, exist *bool) (
 		db.Unlock(kvlock)
 	}()
 
-	initState, err := snapAndReadClusterInfo()
+	clusterInfo, err := readClusterInfo()
 	if err != nil {
 		dlog.Panicln(err)
 	}
-	defer func() {
-		if initState.Collector != nil {
-			initState.Collector.Stop()
-		}
-	}()
 
-	selfNodeEntry, ok := initState.ClusterInfo.NodeEntries[c.config.NodeId]
+	selfNodeEntry, ok := clusterInfo.NodeEntries[c.config.NodeId]
 	if ok && selfNodeEntry.Status == api.Status_STATUS_DECOMMISSION {
 		msg := fmt.Sprintf("Node is in decommision state, Node ID %s.",
 			c.selfNode.Id)
 		dlog.Errorln(msg)
-		return nil, nil, errors.New(msg)
+		return nil, true, nil, errors.New(msg)
 	}
-
-	// Cluster database max size... 0 if unlimited.
-	c.size = initState.ClusterInfo.Size
 	// Set the clusterID in db
-	initState.ClusterInfo.Id = c.config.ClusterId
+	clusterInfo.Id = c.config.ClusterId
+	self, nodeExists := c.initNode(&clusterInfo)
 
-	if initState.ClusterInfo.Status == api.Status_STATUS_INIT {
+	if clusterInfo.Status == api.Status_STATUS_INIT {
 		dlog.Infoln("Initializing a new cluster.")
+		// Initialize self node
+		clusterInfo.Status = api.Status_STATUS_OK
 
-		c.status = api.Status_STATUS_OK
-		initState.ClusterInfo.Status = api.Status_STATUS_OK
-		self, _ := c.initNode(initState.ClusterInfo)
-		err = c.initCluster(initState, self, false)
+		nodeExists = false
+		err = c.initClusterForListeners(self)
 		if err != nil {
 			dlog.Errorln("Failed to initialize the cluster.", err)
-			return nil, nil, err
+			return nil, nodeExists, nil, err
 		}
-	} else if initState.ClusterInfo.Status&api.Status_STATUS_OK > 0 {
-		dlog.Infoln("Cluster state is OK... Joining the cluster.")
-
-		c.status = api.Status_STATUS_OK
-
-		self, exist := c.initNode(initState.ClusterInfo)
-
-		err = c.initNodeInCluster(initState, self, exist)
+		// While we hold the lock write the cluster info
+		// to kvdb.
+		_, err := writeClusterInfo(&clusterInfo)
 		if err != nil {
-			dlog.Errorln("Failed to initialize node in cluster.", err)
-			return nil, nil, err
+			dlog.Errorln("Failed to initialize the cluster.", err)
+			return nil, nodeExists, nil, err
 		}
+	} else if clusterInfo.Status&api.Status_STATUS_OK > 0 {
+		dlog.Infoln("Cluster state is OK... Joining the cluster.")
 	} else {
-		return nil, nil, errors.New("Fatal, Cluster is in an unexpected state.")
+		return nil, nodeExists, nil, errors.New("Fatal, Cluster is in an unexpected state.")
 	}
+	// Cluster database max size... 0 if unlimited.
+	c.size = clusterInfo.Size
+	c.status = api.Status_STATUS_OK
+	return &clusterInfo, nodeExists, self, nil
+}
 
-	selfNodeEntry, ok = initState.ClusterInfo.NodeEntries[c.config.NodeId]
+func (c *ClusterManager) startListeners(db kvdb.Kvdb, nodeExists *bool) (
```

Comment:
let us call this initListeners, to differentiate with actually starting listeners and to clarify what it is doing.